File: entry/entry_state.py
"""Enumeration of lifecycle states for an Entry."""

from enum import Enum, auto


class EntryState(Enum):
    """Lifecycle state of an ``Entry``.

    Members
    -------
    READY
        Data is available and fully materialised.
    POOLED
        Entry is stored in a pool.
    POOLING
        Entry is being transferred to a pool.
    STAGED
        Entry is staged but data may not yet be materialised.
    STALE
        Entry data is out of date.
    NA
        Not applicable. Used by Entry subclasses whose lifecycle states are
        meaningless (e.g. ``Manifest``). Only ``Entry`` may hold non-``NA``
        states.
    """
    READY = auto()
    POOLED = auto()
    POOLING = auto()
    STAGED = auto()
    STALE = auto()
    NA = auto()

File: laila-c/src/entry_state.cpp
#include "laila/entry_state.hpp"

#include "laila/status.hpp"

namespace laila_c {

const char* entry_state_name(EntryState s) {
  switch (s) {
    case EntryState::READY: return "READY";
    case EntryState::STAGED: return "STAGED";
    case EntryState::POOLING: return "POOLING";
    case EntryState::POOLED: return "POOLED";
    case EntryState::STALE: return "STALE";
    case EntryState::NA: return "NA";
  }
  return "STAGED";
}

EntryState entry_state_from_name(const std::string& name) {
  if (name == "READY") return EntryState::READY;
  if (name == "STAGED") return EntryState::STAGED;
  if (name == "POOLING") return EntryState::POOLING;
  if (name == "POOLED") return EntryState::POOLED;
  if (name == "STALE") return EntryState::STALE;
  if (name == "NA") return EntryState::NA;
  return EntryState::STAGED;
}

}  // namespace laila_c
